velox/type/DecimalUtils.h

```diff
@@ -0,0 +1,116 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/LongDecimal.h"
+#include "velox/type/ShortDecimal.h"
+#include "velox/type/Type.h"
+
+namespace facebook::velox {
+static const facebook::velox::int128_t POWERS_OF_TEN[]{
+    1,
+    10,
+    100,
+    1000,
+    10000,
+    100000,
+    1000000,
+    10000000,
+    100000000,
+    1000000000,
+    10000000000,
+    100000000000,
+    1000000000000,
+    10000000000000,
+    100000000000000,
+    1000000000000000,
+    10000000000000000,
+    100000000000000000,
+    1000000000000000000,
+    1000000000000000000 * (int128_t)10,
+    1000000000000000000 * (int128_t)100,
+    1000000000000000000 * (int128_t)1000,
+    1000000000000000000 * (int128_t)10000,
+    1000000000000000000 * (int128_t)100000,
+    1000000000000000000 * (int128_t)1000000,
+    1000000000000000000 * (int128_t)10000000,
+    1000000000000000000 * (int128_t)100000000,
+    1000000000000000000 * (int128_t)1000000000,
+    1000000000000000000 * (int128_t)10000000000,
+    1000000000000000000 * (int128_t)100000000000,
+    1000000000000000000 * (int128_t)1000000000000,
+    1000000000000000000 * (int128_t)10000000000000,
+    1000000000000000000 * (int128_t)100000000000000,
+    1000000000000000000 * (int128_t)1000000000000000,
+    1000000000000000000 * (int128_t)10000000000000000,
+    1000000000000000000 * (int128_t)100000000000000000,
+    1000000000000000000 * (int128_t)1000000000000000000,
+    1000000000000000000 * (int128_t)1000000000000000000 * (int128_t)10,
+    1000000000000000000 * (int128_t)1000000000000000000 * (int128_t)100};
+
+template <typename T>
+std::string decimalToString(const T& value, const TypePtr& type) {
+  VELOX_UNSUPPORTED();
+}
+
+template <>
+std::string decimalToString<LongDecimal>(
+    const LongDecimal& value,
+    const TypePtr& type);
+
+template <>
+std::string decimalToString<ShortDecimal>(
+    const ShortDecimal& value,
+    const TypePtr& type);
+} // namespace facebook::velox
+
+namespace {
+std::string formatDecimal(
+    uint8_t scale,
+    facebook::velox::int128_t unscaledValue) {
+  if (unscaledValue == 0)
+    return "0";
+  std::string result;
+  if (unscaledValue < 0) {
+    result.append("-");
+    unscaledValue = ~unscaledValue + 1;
+  }
+  facebook::velox::int128_t integralPart =
+      unscaledValue / facebook::velox::POWERS_OF_TEN[scale];
```

Comment:
add check for `0 <= scale < sizeof(POWERS_OF_TEN)`

Reply:
Done